### PR TITLE
[registrar] Fix resolving linked away generic types. Fixes #3523.

### LINKED ISSUE #3523: [XI]Building watchintent sample with Release config failed with MTOUCH : error MT4116: Could not register the assembly 'Xamarin.iOS': System.NotSupportedException: Specified method is not supported. "
### Steps to Reproduce

1. Open https://github.com/olegoid/SubmissionSamples/tree/master/WatchOSIntents
2. Build with Rlelease config

### Expected Behavior
app builds successfully

### Actual Behavior
Failed to build with error: " MTOUCH : error MT4116: Could not register the assembly 'Xamarin.iOS': System.NotSupportedException: Specified method is not supported."

### Environment

```
mtouch 11.10.0.1 (d15-7: eadb8376)
```

### Build Logs

https://gist.github.com/GouriKumari/76becf594cc0865a5462cd05aec19a2f

###Regression yes: Built with d15-6

### PR DESCRIPTION
**Pull request by rolfbjarne**
Fixes #3523.

### CODE REVIEW

File: tests/mtouch/RegistrarTest.cs [hunk 1]
```diff
@@ -1414,6 +1414,43 @@ class CTP4 : CTP3 {
 			}
 		}
 
+		[Test]
+		public void LinkedAwayGenericTypeAsOptionalMemberInProtocol ()
+		{
+			// Normally compilable test code can go into monotouch-test, but in this case it depends on the
+			// generic type (NSSet) being linked away, and since monotouch-test uses NSSet, it's not linked
+			// away, and we don't hit the required condition causing the bug.
```
Comment (spouliot) [inline tests/mtouch/RegistrarTest.cs]: maybe `link sdk` or `link all` would have fit the bill ?

Comment (rolfbjarne) [reply tests/mtouch/RegistrarTest.cs]: Good idea, I've put it in `link all` instead.

File: tests/linker/ios/link all/LinkAllTest.cs [hunk 2]
```diff
@@ -574,6 +574,23 @@ void CheckAsyncTaskMethodBuilder (Type atmb)
 			Assert.Null (oifd,  atmb.FullName + ".ObjectIdForDebugger");
 #endif
 		}
+
+		[Test]
+		public void LinkedAwayGenericTypeAsOptionalMemberInProtocol ()
+		{
+			// https://github.com/xamarin/xamarin-macios/issues/3523
+			// This test will fail at build time if it regresses (usually these types of build tests go into monotouch-test, but monotouch-test uses NSSet<T> elsewhere, which this test requires to be linked away).
```
Comment (spouliot) [inline tests/linker/ios/link all/LinkAllTest.cs]: 👍

**Comment by rolfbjarne** [comment]
Test failure is unrelated:

* MMP regression tests: https://github.com/xamarin/maccore/issues/645

**Pull request merged**